cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -543,287 +421,474 @@ private void updateConfigurationTable() {
-          LOG.info("Setting scheduler queue to {} for master services", queueName);
-          preparer.setSchedulerQueue(queueName);
+    /**
+     * Returns the {@link TwillController} for the current master service or {@code null} if none is running.
+     */
+    @Nullable
+    private TwillController getCurrentTwillController(TwillRunnerService twillRunner) {
+      int count = 100;
+      long sleepMs = LOOKUP_ATTEMPT_TIMEOUT_MS / count;
+
+      // Try to lookup the existing twill application
+      for (int i = 0; i < count; i++) {
+        TwillController result = null;
+        for (TwillController controller : twillRunner.lookup(Constants.Service.MASTER_SERVICES)) {
+          if (result != null) {
+            LOG.warn("Stopping one extra instance of {}", Constants.Service.MASTER_SERVICES);
+            try {
+              controller.terminate();
+              controller.awaitTerminated();
+            } catch (ExecutionException e) {
+              LOG.warn("Exception while Stopping one extra instance of {} - {}", Constants.Service.MASTER_SERVICES, e);
+            }
+          } else {
+            result = controller;
+          }
         }
+        if (result != null) {
+          return result;
+        }
+        try {
+          TimeUnit.MILLISECONDS.sleep(sleepMs);
+        } catch (InterruptedException e) {
+          break;
+        }
+      }
+      return null;
+    }
 
-        // Add HBase dependencies
-        preparer.withDependencies(baseInjector.getInstance(HBaseTableUtil.class).getClass());
+    /**
+     * Starts the {@link TwillApplication} for the master services.
+     *
+     * @return The {@link TwillController} for the application.
+     */
+    private TwillController startTwillApplication(TwillRunnerService twillRunner,
+                                                  ServiceStore serviceStore,
+                                                  TokenSecureStoreUpdater secureStoreUpdater) {
+      try {
+        // Create a temp dir for the run to hold temporary files created to run the application
+        Path tempPath = Files.createDirectories(new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                                                         cConf.get(Constants.AppFabric.TEMP_DIR)).toPath());
+        final Path runDir = Files.createTempDirectory(tempPath, "master");
+        try {
+          Path cConfFile = saveCConf(cConf, runDir.resolve("cConf.xml"));
+          Path hConfFile = saveHConf(hConf, runDir.resolve("hConf.xml"));
+          Path logbackFile = saveLogbackConf(runDir.resolve("logback.xml"));
+
+          TwillPreparer preparer = twillRunner.prepare(
+            new MasterTwillApplication(cConf, cConfFile.toFile(), hConfFile.toFile(),
+                                       getSystemServiceInstances(serviceStore)));
+
+          if (cConf.getBoolean(Constants.COLLECT_CONTAINER_LOGS)) {
+            if (LOG instanceof ch.qos.logback.classic.Logger) {
```

Comment:
It was the existing code, I just moved it. I think it's because we are doing the casting inside the `LogHandler` callback.